File: 744-network-delay-time/network-delay-time.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<pair<int, int>>> adj(n + 1); 
        for (auto& it : times) {
            adj[it[0]].push_back({it[1], it[2]});
        }

        vector<int> min_time(n + 1, 1e9);
        min_time[k] = 0;

        queue<pair<int,int>>pq;
        pq.push({0, k});

        while (!pq.empty()) {
            int time = pq.front().first;
            int node = pq.front().second;
            pq.pop();

            for (auto& it : adj[node]) {
                int newNode = it.first;
                int newTime = it.second;

                if (time + newTime < min_time[newNode]) {
                    min_time[newNode] = time + newTime;
                    pq.push({min_time[newNode], newNode});
                }
            }
        }

        int ans = *max_element(min_time.begin() + 1, min_time.end());
        return (ans == 1e9) ? -1 : ans;
    }
};

File: 744-network-delay-time/network-delay-time.py
import heapq
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adj=[[] for _ in range(n+1)]
        for u,v,w in times:
            adj[u].append((v,w))
        
        pq=[]
        heapq.heappush(pq,[0,k])
        min_time=[float('inf')]*(n+1)
        min_time[k]=0
        while pq:
            time,node=heapq.heappop(pq)
            for next_node,next_time in adj[node]:
                if time+next_time<min_time[next_node]:
                    min_time[next_node]=time+next_time
                    heapq.heappush(pq,[time+next_time,next_node])
        max_time = max(min_time[1:])
        
        return max_time if max_time != float('inf') else -1